Don't crash on invalid repoid from yumdb

dmnks opened the pull request:
Ensure the ui_from_repo property always returns ascii-only strings.

Previously, if the yumdb had been tampered with and a repoid was
corrupted for some reason so that it contained non-ascii chars, we could
end up reading it and then crash on an implicit conversion to unicode,
e.g. in output.fmtColumns().

We could also explicitly convert the data passed to fmtColumns() but
that may not fix this completely as we couldn't guarantee that, with the
same corrupted repoid, such an implicit conversion wouldn't happen
elsewhere.  Likewise, doing the conversion in ui_from_repo and returning
a unicode string may cause an implicit conversion elsewhere.

dmnks commented:
Please review the updated version.

dmnks commented:
The reason for the extra commit 09572d3:
I realized that there was a case where the "installed" string would be returned (regardless of the original bug fix) -- and that's when the yumdb has the "from_repo" key unset. So the commit makes the function return "@unknown" instead.

dmnks commented:
OK nevermind.  I removed the extra commit as it would touch quite a few pieces of code in random places so let's stay on the safe side.  It's not needed to fix this bug anyway.

james-antill commented:
invalid_repo is a valid repo id, so don't use that. Probably best to treat it like a from_repo didn't exist, but almost anything that isn't valid is fine.

dmnks commented:
Currently, if from_repo doesn't exist, we print just "installed". Did you mean treating this situation the same way, i.e. printing "installed" in such a case too?

james-antill commented:
Yeh, just printing "installed" is fine.

dmnks commented:
OK, I've done that.

megaumi commented:
> Yeh, just printing "installed" is fine.

Wouldn't that kind of break 'yum list all' output? Since installed-from-repo is supposed to be prefixed by "@"

megaumi commented:
And, well, "installed" is also a valid repo name.

dmnks commented:
Valentina is right. Whatever we print in the repo column should be an actual repo name (according to the docs). Which is currently not the case if from_repo is undefined in yumdb for the given package.

So the question is: what should we print if the repo name is unkown or invalid? Two solutions come to mind:
- choose a reserved repo name that would act as a placeholder for such cases (and make that particular repo name also invalid), or
- don't print anything

megaumi commented:
> choose a reserved repo name that would act as a placeholder for such cases (and make that particular repo name also invalid)

I'd go with this approach, and also add it to the man page.

james-antill commented:
> Wouldn't that kind of break 'yum list all' output? Since installed-from-repo is supposed to be prefixed by "@"

I don't understand, we can already print "installed" if from_repo doesn't exist. This just extends that to "doesn't exist or is invalid", which seems like a fine extension. So the patch   4f55d4d LGTM +1.

Using some other value is ok, if we want the user to do something with that information. The problem is that we'd have to include one of the forbidden characters to create a "name that can't be found in the wild" and I'm not sure what the user is supposed to do (given the common case is corruption, the obvious solution would be to just delete from_repo, which would then revert the name to "installed" anyway so why not do it automatically?).
In theory we could match the checksum_data to packages that are available, and if we find a match reset from_repo (and/or look in history for the from_repo, and even maybe just take a guess based on NEVRA). At which point it makes more sense to print "@run:yum-check-from_repos" or something.

megaumi commented:
> I don't understand, we can already print "installed" if from_repo doesn't exist. This just extends that to "doesn't exist or is invalid", which seems like a fine extension.

Right, it's not breaking anything, but it's extending the confusing undocumented behavior.

Maybe handling this corner case is not worth changing this behavior, but we should, at least, have it documented.

dmnks commented:
I'm working on a better version of this patch as I found that the proposed one doesn't fix the bug entirely.

dmnks committed:
Allow for validating attributes read from yumdb

Make sure we don't expose corrupted attributes read from the yumdb to
the consumers.  There's at least one report of such a corruption: BZ
1234967.  Instead, make requesting a malformed yumdb attribute
equivalent to requesting a non-existent one -- which is a valid
scenario, already handled by the consumers.

Note that the actual validator function that fixes the above bug will be
committed separately.

dmnks committed:
Don't crash on invalid from_repo in yumdb. BZ 1234967

Implement a yumdb validator function for the from_repo attribute.  This
prevents yum from crashing if an implicit conversion to unicode takes
place somewhere and the attribute contains non-ascii chars due to some
yumdb corruption.

Reproducers:

$ yum install foo
$ yumdb set from_repo <non-ascii-chars> foo
$ yum list foo  # crash
$ yum --disablerepo=<repo-with-foo> reinstall foo  # crash
$ yum --verbose version installed  # crash

dmnks committed:
docs: mention special case for unknown from_repo

dmnks commented:
Reworked version pushed.
